Implements #304: Adds deprecated-bugs.csv and refactors commit-db into active-bugs.csv

Greg4cr opened the pull request:
This PR fixes #304. 

* `commit-db` has been refactored into `active-bugs.csv`.
* A header has been added to the first line of `active-bugs.csv` noting what each column represents.
* All code handling `active-bugs.csv` has been adjusted to handle the header.
* `deprecated-bugs.csv` has been added, containing the entries for all deprecated bugs.
* `deprecated-bugs.csv` utilizes the same header as `active-bugs.csv`, with the addition of the D4J version where the bugs was deprecated and the reason for deprecation.
* `Constants.pm` has been updated with constants representing the column names.

rjust commented:
@Greg4cr, thanks for implementing this. Can you please define the file names in Constants.pm as well and make sure that every script uses the constants? Linking to these constants in the pod (as opposed to using the file name) would also be good to avoid inconsistencies in case these file names change.

Greg4cr commented:
@rene - 

I have pushed commits that:
* Add constants for the filenames, and use that constant elsewhere. 
* Renamed the column names in the CSV files to match style and usage elsewhere in the framework.
* Refactored `test_export_command` to use the method `get_bug_ids` instead of duplicating the code in that method. 

I looked at the use of system calls to tail in the bug mining framework. There is only **one** use of tail in the bug mining framework, in `initialize-project-and-collect-issues.pl`. This is when we mine more bugs for an existing project, and are filtering commit hashes that already appear in the existing project. I don't think this is worth refactoring right now. I don't know if there is a good way to surface that information via DBI, and it represents a fairly specialized corner case. This is something we may want to file a bug report for and tackle as part of a broader clean-up of the bug mining code.

jose commented:
Thanks @Greg4cr for implementing this.

1. I'm wondering whether we want to rename `commit-db` to something else.  This single change will break **all** my scripts and I believe others' as well.
2. The brand new `active-bugs.csv` and `deprecated-bugs.csv` do include a header line, which will also break **all** my scripts (even after addressing 1).

How about we keep `commit-db` as it is and add an extra column `status` which could be `active` or `deprecated`?

jose commented:
By the way, @Greg4cr, I noticed in a couple pull requests that you've been mentioning @rene instead of @rjust.   :-)

Greg4cr commented:
@jose Haha, oops! I should proofread the comments more thoroughly.

This change came out of discussions @rjust and I had earlier this week. We discussed the concerns you mentioned. At the time, this is what we decided on:
* `active_bugs.csv` and `deprecated_bugs.csv` makes a clear distinction between current and deprecated metadata that `commit-db` does not.
* Two files instead of one enforces separation between bugs that should be used and those that are no longer considered valid - but that we retain information on for historical reasons.
* A header should be added to document what these fields actually represent. This helps reduce confusion and questions when using the framework.

Something that Rene and I discussed, and that I certain agree with, is that **internal** changes to Defects4J can and should go forward even if it would impact **external** scripts. Even if we didn't make this particular change, other changes to Defects4J will break someone's scripts. We shouldn't let that prevent changes to the framework.

I'm certainly willing to discuss alternative implementations, though. @rjust - thoughts?

rjust commented:
I have strong opinions about improving the internals of the framework such as separating active and deprecated bugs, consistent use of headers, consistent naming, removing code clones, etc., but I do agree that backward compatibility is a concern to keep in mind. IMO, the problem with commit-db (and why this file is used in so many wrapper scripts) is the lack of a simple, public API to obtain the list of valid project ids and bug ids -- we already floated some ideas for addressing this issue.

I am OK with leaving the commit-db file in the repo for now and clearly documenting that (1) it is deprecated and (2) will be removed in the future. At the same time, we should offer a clear alternative, which is not use the new file name instead and skip the header. If we want others to migrate away from using commit-db, which is not part of the public API, we should make sure that there is an alternative public API. Iterating over all bug ids should be as simple as `for bid in $(defects4j <info, export, or new cmd> -p $pid)`. Note that v2.0.0 will break existing scripts that assume that bug ids correspond to line numbers in commit-db. Having documented a preferred way of obtaining valid project and bug ids would clearly help users.

rjust commented on framework/core/Constants.pm:
```diff
@@ -256,6 +256,22 @@ our $TAG_FIXED            = "FIXED_VERSION";
+our $BUGS_CSV_DEPRECATED_WHY = "deprecated_reason";
+
+# Reasons for deprecation
+our $DEPRECATED_JVM8 = "JVM8-NonReproducable";
```
There is another reason for deprecation: a few bugs were duplicates. Also, it might be worth separating non-reproducible from non-compilable bugs. Reproducable should be Reproducible.

Greg4cr replied:
I have updated and added the following:
`our $DEPRECATED_DUPLICATE = "Duplicate";`
`our $DEPRECATED_JVM8_REPRO = "JVM8_Not_Reproducible";`
`our $DEPRECATED_JVM8_COMPILE = "JVM8_Does_Not_Compile";`

I changed Closure 63 and 93 to `Duplicate`. Were there any other duplicates?

rjust commented on framework/core/Vcs.pm:
```diff
@@ -115,15 +115,15 @@ A Vcs object has to be instantiated with:
+=head2 active-bugs csv
 
-The commit-db (csv) file has the structure: C<bug_id,revision_buggy,revision_fixed>.
+The active-bugs (csv) file has the structure: C<bug_id,revision_buggy,revision_fixed>.
```
This documentation is outdated. There are more fields in this file.

Greg4cr replied:
Updated

rjust commented on framework/core/Vcs.pm:
```diff
@@ -231,7 +231,7 @@ sub contains_version_id {
-Performs a lookup of C<vid> in the C<commit-db> followed by a checkout of
+Performs a lookup of C<vid> in the C<active-bugs csv> followed by a checkout of
 the corresponding revision with C<revision_id> to F<work_dir>.
 Format of C<vid>: C<\d+[bf]>.
```
Can you replace the "Format of vid ..." occurrences with a pointer to the Utils subroutine that checks for well-formedness?

Greg4cr replied:
In the documentation? Should I replace `Format of C<vid>: C<\d+[bf]>` with `Format of C<vid> is checked using C<Utills::check_vid($vid)>`?

Greg4cr replied:
Links added to Utils and Constants in the documentation.

rjust commented on framework/core/Vcs.pm:
```diff
@@ -409,15 +409,17 @@ sub rev_date {
     while (<IN>) {
         chomp;
-        /(\d+),([^,]+),([^,]+)/ or die "Corrupted commit-db!";
+        /(\d+),([^,]+),([^,]+)/ or die "Corrupted active-bugs csv!";
```
While this does match, it is not corresponding to the expected format.

Greg4cr replied:
Updated

rjust commented on framework/test/test_bug_mining.sh:
```diff
@@ -358,8 +359,9 @@ test_promote_to_db() {
-    [ -s "$HERE/../projects/$project_id/dir-layout.csv" ] || die "dir-layout.csv does not exit or it is empty"
+    [ -s "$HERE/../projects/$project_id/$BUGS_CSV_ACTIVE" ] || die "active-bugs csv does not exist or it is empty"
+    [ -s "$HERE/../projects/$project_id/$BUGS_CSV_DEPRECATED" ] || die "deprecated-bugs csv does not exist or it is missing the header"
+    [ -s "$HERE/../projects/$project_id/dir-layout.csv" ] || die "dir-layout.csv does not exist or it is empty"
```
Would be good to have a constant for this file as well -- maybe in a dedicated PR that cleans up the metadata (see 
also #309).

Greg4cr replied:
For `dir_layout.csv`? I can add one if you'd like.

Greg4cr replied:
Added constant for `dir_layout.csv` in latest commit.

rjust commented on framework/test/test_gen_tests.sh:
```diff
@@ -60,7 +60,7 @@ init
 if [ "$BUGS" == "" ]; then
-    num_bugs=$(num_lines $BASE_DIR/framework/projects/$PID/commit-db)
+    num_bugs=$(num_lines $BASE_DIR/framework/projects/$PID/$BUGS_CSV_ACTIVE)
     BUGS="$(seq 1 1 $num_bugs)"
```
This should be replaced with a call to get_bug_ids. This also makes the check on line 86 obsolete.

Greg4cr replied:
Fixed

rjust commented on framework/test/test_verify_bugs.sh:
```diff
@@ -93,9 +93,9 @@ work_dir="$test_dir/$PID"
-        warn "Skipping bug ID that is not listed in commit-db: $PID-$bid"
+    # Skip all bug ids that do not exist in the active-bugs csv
+    if ! grep -q "^$bid," "$BASE_DIR/framework/projects/$PID/$BUGS_CSV_ACTIVE"; then
+        warn "Skipping bug ID that is not listed in active-bugs csv: $PID-$bid"
```
Seems unnecessary, given that the bug ids are extracted from the same file on line 71. Delete this check?

Greg4cr replied:
This check is needed if the user specified a list of bugs that don't exist. Line 71 is only run if the user does not already specify a list of bugs.

Greg4cr commented:
@rjust @jose 

>> I am OK with leaving the commit-db file in the repo for now and clearly documenting that (1) it is deprecated and (2) will be removed in the future.

I agree. We can include the commit-db as well, with a disclaimer that it will be deprecated and removed in the future.

>> At the same time, we should offer a clear alternative, which is not use the new file name instead and skip the header. If we want others to migrate away from using commit-db, which is not part of the public API, we should make sure that there is an alternative public API. Iterating over all bug ids should be as simple as for bid in $(defects4j <info, export, or new cmd> -p $pid). 

I also agree with this. I was initially thinking about this as an extension of defects4j info, but it might be clumsy to build it on top of that existing command. Perhaps it would be better to have a new command built around exporting info from active/deprecated bugs files:
`defects4j bug-query -p $pid -q $query`
(or some other name)
* Basic call (just project, no query) would print the entire `active-bugs.csv` for a project
* `$query` is a comma separated list of fields. These must correspond to the field names defined in the CSV header. This allows definition of a custom set of fields to be returned
  * For example, `"bug_id"` would return just bug IDs, `"bug_id,bug_report_id"` would return bug IDs and issue tracker IDs.
* Additional flags then control active/deprecated inclusion:\
  * No flag prints info on active bugs only.
  * `-D` prints info on deprecated bugs only.
  * `-A` returns both active and deprecated bugs.
* `-H` includes a header in what is printed. By default, a header will not be printed.

Greg4cr commented:
I added another commit making minor tweaks (see @rjust's questions above) and restoring the commit-db files (with the intention that they are deprecated and will later be phased out, and that the framework makes use instead of `active-bugs.csv`).

jose commented:
@Greg4cr, I would keep it simple and try to implement this as an extension of the existing `defects4j info`, instead of yet another command.

`defects4j info -p $pid` already returns the `Number of bugs:`. I would just augment the output to include something like:
```
Active bugs:1,2,3,6
Deprecated bugs:4,5
```
So that anyone could do `defects4j info -p $pid | grep "^Active bugs: " | cut -f2 -d':'` and have the list of bug ids.

Greg4cr commented:
Meeting discussion today:
* Implement API as a new command, `d4j-query` (or some other name)
  * I will think about semantics of query. 
  * Base idea - provide Project and a comma-separated set of fields, get metadata info that is applicable. See comment above.
* Extend to other metadata potentially too, not just active/deprecated bugs. 
  * Add constants to `core/Constants.pm` for any additional fields

Greg4cr commented:
@rjust @jose 

I have implemented the `d4j-query` command, which can be used to query the project metadata for particular fields. Some examples:

`defects4j query -p Collections -h` returns the list of available fields that can be queried.
`defects4j query -p Collections` returns the list of active bug IDs for project Cli.
`defects4j query -p Collections -q "revision_id_buggy,modified_classes"` returns all active bug IDs, and for each, the revision hash of the buggy version and the list of modified classes.
`defects4j query -p Collections -q "revision_id_buggy,modified_classes"` returns all deprecated bug IDs, and for each, the revision hash of the buggy version and the list of modified classes.
`defects4j query -p Collections -q "revision_id_buggy,modified_classes" -a> returns all bug IDs (active and deprecated), and for each, the revision hash of the buggy version and the list of modified classes.
`defects4j query -p Collections -q "deprecated_reason" -a` returns all bug IDs (active and deprecated) along with the reason for deprecation. For active bugs, the deprecation reason will be "NA", as those bugs do not have values for that field.

`test_d4j_query.sh` demonstrates those examples (with expected output in `framework/test/resources/output/d4j-query/`). 

Feel free to try it out and suggest improvements.

Greg4cr commented:
@rjust @jose - Any thoughts or issues on this? Can we merge, or would you like any changes to the `d4j-query` command?

Greg4cr commented:
@rjust @jose 

One thing this PR does not address right now is check out of deprecated bugs
1) Should we allow check out at all of deprecated bugs?
2) If yes, how should we handle this? Options include:
  - Warn on checkout, but allow everything
  - Warn on checkout, disable `d4j-test` (and `d4j-compile` for uncompilable bugs)

bkushigian commented:
@Greg4cr: @rjust and I were chatting about this feature yesterday during our meeting and here's a quick summary: Rene was thinking long term about a good organization for the different commands. Rene's concern is that there will be four different commands that kind of step on each others toes: `info`, `query`, `export`, and `env`. These each report some sort of information about defects4j's database and environment. I don't totally grok the differences between `export` and `info` but from what I understand there are some implementation-related things that make it useful to separate them (stuff about having to check out a project for export-related commands).

Anyway, here are a few thoughts on possible unifying approaches.
1. Make `info` more robust and include `query` in `info`
2. Make the differences between `query` and `info` explicit. Currently `info` prints more human-facing info, while `query` looks to be geared towards a lower level representation. Maybe it makes sense to make `query` the plumbing for d4j, and make `info` the porcelain? If this happens, it might be worth while to refactor info to rely on `query`. There will still be some decisions to make---`info` prints stuff about where scripts are located, etc. Is this a `query` question or an `env` question? Should `env` provide a command for script locations?
3. Long term we might also want to support some sort of aliasing for sub commands. For instance, from where I am as an end user, I'd love to be able to run `defects4j bugs -pLang` and get the current output of `defects4j query -pLang`. Having nice high level aliases to complex functionality might be useful if `query` is going to be our main data-querying plumbing.

But yeah, definitely looking forward to getting this into `master` so I can use it!

@rjust  any followups/things that need doing?

bkushigian commented:
And in response to @jose's comment about about printing `Active bugs:1,3,4,5,...` and grepping, maybe this is the way. As an end user I'd rather be able to just run `defects4j bugs -pLang` and not have to worry about `grep`/`cut` for two reasons:
1. it's easier to write
2. it's easier to read
3. I'd rather not deal with raw text semantically---if I have to parse user-facing human-readable stuff that isn't in a structured and well defined format I'm not going to be sure that it won't change on me. Will there be a commit that changes formatting `Active bugs:` -> `Active Bugs:`?

Conversely, adding more commands makes things more complicated and adds extra work on the development end.

Greg4cr commented:
@bkushigian @rjust @jose 

> Rene's concern is that there will be four different commands that kind of step on each others toes: `info`, `query`, `export`, and `env`

This is a reasonable concern, and one that Rene and I have also discussed. There are multiple ways this could be addressed.

The intent of `info` is a human-facing interface and the intent of `query` is a scripting interface, but both access project-related information. These could be combined (i.e., using a flag to control whether output is pretty or not). Alternatively, it may make sense to keep them separate and make the separate intents clear.

`export` is specifically a way to extract Ant properties, if I recall correctly. It is kind of its own thing, and I think should be kept separate. It is also version-specific, not project or bug-specific. It operates at a different level of granularity.

`env` is for examining the value of environment variables. Again, I see this as a separate utility altogether. `query`and `info` are specifically for project metadata. I don't see environment variables as part of the same use case. Again, a different level of granularity. 

So, in terms of combining tools, I see a case for merging `info` and `query`, but not the other two. I also see a case for keeping `info` and `query` separate, and making the differing intent explicit. Let's discuss and make some decisions.

(as a TODO for myself, `info` does print some information that is not currently offered by `query`. When I have some time, I'll see what other information it would make sense for `query` to surface from the metadata.)

bkushigian commented:
This makes sense.

Greg4cr commented:
@rjust @jose 

Some updates:

- `d4j-query` now can surface several additional pieces of information from the metadata. A full list of fields is documented in: https://github.com/rjust/defects4j/blob/bugs-csv/README.md#export-project-specific-metadata 
- I've added a flag that can output the extracted csv to a file 
- I added a high-level command `d4j-bugs` that just prints the list of bug IDs.

rjust commented on README.md:
```diff
@@ -159,6 +169,48 @@ directory to export a version-specific property:
+| project.name          | Original repository name                                                            |
+| revision.id.buggy     | Commit hashes for the buggy version of each bug                                     |
+| revision.id.fixed     | Commit hashes for the fixed version of each bug                                     |
+| revision.buggy.date   | Date of the buggy commit for each bug                                               |
+| revision.fixed.date   | Date of the fixed commit for each bug                                               |
```
To be consistent with the above two properties, this should be revision.date.<buggy|fixed>

Greg4cr replied:
Corrected in latest commit.

rjust commented on README.md:
```diff
@@ -159,6 +169,48 @@ directory to export a version-specific property:
+| revision.fixed.date   | Date of the fixed commit for each bug                                               |
+| report.id             | Bug report ID from the version tracker for each bug                                 |
+| report.url            | Bug report URL from the version tracker for each bug                                |
+| build.file            | Location of the Defects4J build file for the project                                |
```
Should this be project.build.file?

Greg4cr replied:
Corrected in latest commit.

rjust commented on README.md:
```diff
@@ -159,6 +169,48 @@ directory to export a version-specific property:
+| report.id             | Bug report ID from the version tracker for each bug                                 |
+| report.url            | Bug report URL from the version tracker for each bug                                |
+| build.file            | Location of the Defects4J build file for the project                                |
+| vcs                   | Version control system used by the project                                          |
+| repository            | Location of the project repository                                                  |
+| commit.db             | Location of the CSV containing information on that bug                              |
```
Similar to build.file, should these be prefixed with project?

Greg4cr replied:
I agree that it would be more consistent. Corrected in latest commit.

rjust commented on README.md:
```diff
@@ -159,6 +169,48 @@ directory to export a version-specific property:
+| deprecated.version    | (for deprecated bugs only) Version of Defects4J where a bug was deprecated          |
+| deprecated.reason     | (for deprecated bugs only) Reason for deprecation                                   |
+
+By default, `defects4j query` returns information on active bugs. The `[-d]`
+flag returns information only on deprecated bugs, while the `[-a]` flag returns
```
Usually, flags are lower-case if they take an argument and upper-case if they don't. I think these should be -D and -A.

Greg4cr replied:
Corrected in latest commit.

rjust commented on framework/bin/d4j/d4j-bugs:
```diff
@@ -0,0 +1,99 @@
+
+=head1 SYNOPSIS
+
+  d4j-bugs -p project_id [-d] [-a]
```
See comment above: these should be -D and -A.

Greg4cr replied:
Corrected in latest commit.

rjust commented on framework/bin/d4j/d4j-bugs:
```diff
@@ -0,0 +1,99 @@
+
+=head1 NAME
+
+d4j-bugs -- Lists all bug IDs for a project
```
Should we name this command print-bugs? It's slightly longer but more descriptive. I also suggested to rename env to print-env -- so, we should make sure these are consistent, regardless of which solution we adopt.

Greg4cr replied:
I agree with this. Renamed in latest commit.

rjust commented on framework/bin/d4j/d4j-bugs:
```diff
@@ -0,0 +1,99 @@
+my $ALL_BUGS = defined $cmd_opts{a} ? 1 : 0;
+
+if ($ONLY_DEP and $ALL_BUGS) {
+    die "Only dependencies (-d) and all bugs (-a) cannot be concurrently set.";
```
dependencies -> deprecated

Greg4cr replied:
Corrected in latest commit.

rjust commented on framework/bin/d4j/d4j-bugs:
```diff
@@ -0,0 +1,99 @@
+my $ONLY_DEP = defined $cmd_opts{d} ? 1 : 0;
+my $ALL_BUGS = defined $cmd_opts{a} ? 1 : 0;
+
+if ($ONLY_DEP and $ALL_BUGS) {
```
This should be indicated in the perldoc and usage message: [-D | -A]

Greg4cr replied:
This is now indicated in the documentation.

rjust commented on framework/bin/d4j/d4j-bugs:
```diff
@@ -0,0 +1,99 @@
+    die "Only dependencies (-d) and all bugs (-a) cannot be concurrently set.";
+}
+
+if ($ALL_BUGS) {
+    system('../defects4j','query','-p',$PID,'-a');
+} elsif ($ONLY_DEP) {
+    system('../defects4j','query','-p',$PID,'-d');
+} else {
+    system('../defects4j','query','-p',$PID);
+}
```
This is very brittle. Can you replace the system calls with with calls to an internal query API to obtain and print the bug list?

Greg4cr replied:
Good call. I have created `Query.pm` in `core`, which has the query functionality and an API for accessing it. `d4j-query` and `d4j-print-bugs` now make use of that API.

rjust commented on README.md:
```diff
@@ -137,7 +145,9 @@ Use [`framework/bin/defects4j`](http://defects4j.org/html_doc/defects4j.html) to
 | [mutation](http://defects4j.org/html_doc/d4j/d4j-mutation.html)           | Run mutation analysis on a buggy or a fixed project version                                       |
 | [coverage](http://defects4j.org/html_doc/d4j/d4j-coverage.html)           | Run code coverage analysis on a buggy or a fixed project version                                  |
 | [monitor.test](http://defects4j.org/html_doc/d4j/d4j-monitor.test.html)   | Monitor the class loader during the execution of a single test or a test suite                    |
+| [bids](http://defects4j.org/html_doc/d4j/d4j-bids.html)                   | Print the list of active bug IDs for a specific project                                           |
```
Print the list of active or deprecated bug IDs for a specific project

Greg4cr replied:
Fixed in latest commit

rjust commented on README.md:
```diff
@@ -159,6 +169,48 @@ directory to export a version-specific property:
+|-----------------------|-------------------------------------------------------------------------------------|
+| bug.id                | Assigned bug IDs (included in all results)                                          |
+| project.id            | Assigned project ID                                                                 |
+| project.name          | Original repository name                                                            |
```
Should this be "Original project name"?

Greg4cr replied:
Fixed in latest commit

rjust commented on README.md:
```diff
@@ -159,6 +169,48 @@ directory to export a version-specific property:
+| project.build.file    | Location of the Defects4J build file for the project                                |
+| project.vcs           | Version control system used by the project                                          |
+| project.repository    | Location of the project repository                                                  |
+| project.commit.db     | Location of the CSV containing information on that bug                              |
```
Should this be the field for the active-bugs.csv? Are there now two fields for active and deprecated bugs?

Greg4cr replied:
Changed to project.bugs.csv in latest commit
